### DISCUSSION TITLE
server backend init on startup

### DISCUSSION THREAD

**Discussion by MarcTroll** [nuxt/nuxt#16228] [Questions] [answered]
In my project I want to add some backend logic, that has (almost) nothing to do with nuxt. This includes managing database connections, cronjobs, mail-sending, or something like this.

To do that I created the folder `server/lib`, where my backend logic should be placed. For testing, I only added a single class that has a property that is randomly generated.

In an API-route (`server/api/test.ts`) I now import that class and return the value of the random property. But the problem is, that the class is being loaded when navigating to `localhost:3000/api/test` in my browser. Is there a possibility to initialize such classes when starting the backend server to already connect to the database for example?

Establishing a database connection on the first request requiring a database connection makes the user wait a longer time.

EDIT: Is it intended to do such things in nuxt? If not, what solution is intended for my problem? I won't be the first one who uses databases or mail sending together with nuxt.

**Comment by wobsoriano** [comment]
I think you can create a server plugin for that.

```ts
// plugins/init.server.ts
import { defineNuxtPlugin } from '#app'

export default defineNuxtPlugin(() => {
  // Initiate your database, cronjobs, mail-sending, or whatever here.
})
```

EDIT: Do it in a module instead

```ts
// modules/init.ts
import { defineNuxtModule } from '@nuxt/kit'

export default defineNuxtModule({
    setup() {
        console.log('INIT')
    }
})
```

```ts
// nuxt.config.ts
import { defineNuxtConfig } from 'nuxt3'

// https://v3.nuxtjs.org/docs/directory-structure/nuxt.config
export default defineNuxtConfig({
    buildModules: ['~/modules/init']
})
```

**Reply by MarcTroll** [reply]
Thanks for the reply @wobsoriano . However, when doing something like this, nothing happens on startup:
```typescript
// plugins/init.server.ts
import { defineNuxtPlugin } from "#app";

export default defineNuxtPlugin(() => {
    console.log("INIT!");
});
```

There is no such `INIT!` printed during startup. If searching the `.output`-folder for the output, I can find the plugin in the `server.mjs`, but it is not executed.

**Reply by wobsoriano** [reply]
Are you sure? Check your terminal for the output. If you think it's not working, a simple reproduction would do. But it should be working.

**Reply by MarcTroll** [reply]
When starting the nuxt-app using `npm run start` after `npm run build` no output but the following is printed:
```
npm run start

> start
> node .output/server/index.mjs

Listening on http://localhost:3000
```

When entering any page of the nuxt-application (e.g. `http://localhost:3000`), then the expected output is printed. But then its too late again, because the request is already happening and the user has to wait again for the backend to initialize.

Because codesandbox runs in dev-mode and I can't figure out how to change it: To reproduce this you only have to create a server plugin as in https://github.com/nuxt/framework/discussions/1788#discussioncomment-1611586, build the app (`npm run build`), and start it (`npm run start`). As long as you don't access the nuxt app on the client, the `INIT!` won't be printed.

**Reply by wobsoriano** [reply]
Updated my answer. You can use a module instead.

**Reply by MarcTroll** [reply]
I tried that out, but it seems, they are executed during build time and not during runtime:
```
npm run build

> build
> nuxt build

Nuxt CLI v3.0.0-27272804.ca557e7                                                                                                                                                                                     14:05:52
INIT!                                                                                                                                                                                                                14:05:53
i Vite warmed up in 1445ms                                                                                                                                                                                           14:05:55
i Client built in 1482ms                                                                                                                                                                                             14:05:55
i Building server...                                                                                                                                                                                                 14:05:55
√ Server built in 539ms                                                                                                                                                                                              14:05:55
i Nitro preset is server                                                                                                                                                                                             14:05:55
i Cleaning up .output                                                                                                                                                                                                14:05:55
start Generating public...                                                                                                                                                                                           14:05:55
√ Generated public .output/public                                                                                                                                                                                    14:05:56
start Building server...                                                                                                                                                                                             14:05:56
start Writing server bundle...                                                                                                                                                                                       14:05:57
√ Server built                                                                                                                                                                                                       14:05:58
  ├─ .output/server/index.mjs (69.8 kB) (25.6 kB gzip)
  ├─ .output/server/chunks/vue3.mjs (283 B) (197 B gzip)
  ├─ .output/server/chunks/test.mjs (319 B) (218 B gzip)
  ├─ .output/server/chunks/static.mjs (4.88 kB) (1.56 kB gzip)
  ├─ .output/server/chunks/server.mjs (167 kB) (40.2 kB gzip)
  ├─ .output/server/chunks/render.mjs (22.5 kB) (6.24 kB gzip)
  ├─ .output/server/chunks/index.mjs (390 kB) (89.9 kB gzip)
  ├─ .output/server/chunks/hello.mjs (349 B) (226 B gzip)
  ├─ .output/server/chunks/client.manifest.mjs (1.35 kB) (375 B gzip)
  └─ .output/server/chunks/BookingLib.mjs (219 B) (159 B gzip)
Σ Total size: 999 kB (247 kB gzip)
√ Ready to run node .output/server/index.mjs
```
Here "INIT!" is printed directly at the beginning of executing build script, but is never printed after running `npm run start`.

In https://v3.nuxtjs.org/docs/directory-structure/nuxt.config#buildmodules:
> Modules that are only required during development and build time.

I also tried to use a "normal" [module](https://v3.nuxtjs.org/docs/directory-structure/nuxt.config#modules), but without any success. Same result as using `buildModules` option.

**Comment by MarcTroll** [comment, marked as answer]
@wobsoriano I'm currently trying to do that using a custom server preset. Following the [documentation](https://v3.nuxtjs.org/docs/deployment/presets/custom) leads me to a custom preset:
```typescript
import type {NitroContext, NitroPreset} from '@nuxt/nitro'

const preset: NitroPreset = {
    entry: 'server/entry',
    externals: true,
    serveStatic: true,
    hooks: {
        'nitro:compiled' ({ output }: NitroContext) {
            console.log('Ready to run "node ' + output.serverDir + '/index.mjs"')
        }
    }
}

export default preset
```

After defining the preset I copied the entry-point from the existing entry point for the server-preset from here: https://github.com/nuxt/framework/blob/main/packages/nitro/src/runtime/entries/server.ts

But now I'm wondering how to import the `handle`-function defined in `nuxt/nitro/runtime/server`. Currently I have the following for testing purposes:
```typescript
import '#polyfill'
import { Server } from 'http'
import destr from 'destr'
import { handle } from '@nuxt/nitro' // how to import that here?

const server = new Server(handle)

const port = (destr(process.env.NUXT_PORT || process.env.PORT) || 3000) as number
const hostname = process.env.NUXT_HOST || process.env.HOST || 'localhost'

// @ts-ignore
server.listen(port, hostname, (err) => {
    if (err) {
        console.error(err)
        process.exit(1)
    }
    console.log(`Listening on http://${hostname}:${port}`)
})

export default {}
```

I also tried to import the `handle`-function (line 4) from `@nuxt/nitro/runtime/server` but without any success.
How could I import that function. I would need to use the default implementation, but as before I want to do some prestart-checks/-routines before `server.listen` is called.

**Reply by MarcTroll** [reply]
Ok, got it. Need to import from `@nuxt/nitro/dist/runtime/server`.

**Reply by lewebsimple** [reply]
I'm trying to setup [BullMQ sandboxed processors](https://docs.bullmq.io/guide/workers/sandboxed-processors) in the same project as my Nuxt3 application, and @ilouHD's answer got me closer to a solution.

Does Nitro allow a way to compile worker entries in addition to the single server entry ? i.e. I would like to compile `server/entry.ts` as well as `server/workers/*.ts` so I could tell BullMQ where to find the worker entries in `.output`

Maybe Nitropack will support that... perhaps @pi0 or @danielroe could answer that ?

**Reply by lewebsimple** [reply]
_Answering my own question_: It doesn't seem possible with the current nitro package, as the `getRollupConfig` uses the following:
```
// ...
  const rollupConfig: RollupConfig = {
    input: resolvePath(nitroContext, nitroContext.entry),
    output: {
      dir: nitroContext.output.serverDir,
      entryFileNames: 'index.mjs',
// ...
```

**Reply by zhuscat** [reply]
It seems that custom preset only works in production mode @ilouHD 

So, I think using custom preset is not proper for this use case?

**Reply by maicss** [reply]
See what @danielroe replied https://github.com/nuxt/nuxt/discussions/24998

cc @andysay

**Comment by andysay** [comment]
so how use that in nuxt 3 tooday?

**Reply by wobsoriano** [reply]
What about a Nitro plugin? Something like

```ts
// nitro/cron.ts
import cron from 'node-cron'

export default defineNitroPlugin(() => {
  cron.schedule("*/2 * * * * *", function() {
    console.log("running a task every 2 seconds")
  })
})
```

```ts
export default defineNuxtConfig({
  nitro: {
    plugins: ['@/nitro/cron']
  }
})
```

**Reply by andysay** [reply]
> 

is it possible to access useNuxtApp from nitro plugins?

**Reply by danielroe** [reply]
No.

**Reply by andysay** [reply]
> No.

it may become available in the future ?

**Reply by TheAlexLichter** [reply]
The nuxt app will not be available in a Nitro plugin because the Nitro plugin runs *before* nuxt is even initialized ☺️
